pkg/cri/resource-manager/policy/builtin/podpools/metrics.go

```diff
@@ -0,0 +1,242 @@
+			}
+			sort.Sort(sort.StringSlice(podIds))
+			for _, podId := range podIds {
+				for _, containerId := range pool.PodIDs[podId] {
+					if container, ok := p.cch.LookupContainer(containerId); ok && container != nil {
+						containerName := container.PrettyName()
+						if policyMetrics.PoolMetrics[pool.PrettyName()].ContainerNames == "" {
+							policyMetrics.PoolMetrics[pool.PrettyName()].ContainerNames = containerName
+						} else {
+							policyMetrics.PoolMetrics[pool.PrettyName()].ContainerNames = fmt.Sprintf("%s,%s", policyMetrics.PoolMetrics[pool.PrettyName()].ContainerNames, containerName)
+						}
+					}
+				}
+				if pod, ok := p.cch.LookupPod(podId); ok && pod != nil {
+					podName := pod.GetName()
+					if policyMetrics.PoolMetrics[pool.PrettyName()].PodNames == "" {
+						policyMetrics.PoolMetrics[pool.PrettyName()].PodNames = podName
+					} else {
+						policyMetrics.PoolMetrics[pool.PrettyName()].PodNames = fmt.Sprintf("%s,%s", policyMetrics.PoolMetrics[pool.PrettyName()].PodNames, podName)
+					}
+				}
+			}
+		}
+	}
+	return policyMetrics
+}
+
+// CollectMetrics generates prometheus metrics from cached/polled policys-specific metrics data.
+func (p *podpools) CollectMetrics(m policy.Metrics) ([]prometheus.Metric, error) {
+	metrics, ok := m.(*Metrics)
+	if !ok {
+		return nil, fmt.Errorf("Wrong podpools metrics.")
+	}
+	err := cpuTimeStat.GetCPUTimeStat()
+	if err != nil {
+		return nil, err
+	}
+	threadCPUMetrics := updateThreadCPUUsageMetrics()
+	poolCPUMetrics, err := updatePoolCPUUsageMetrics(metrics)
+	if err != nil {
+		return nil, err
+	}
+	return append(threadCPUMetrics, poolCPUMetrics...), nil
+}
+
+// updateThreadCPUUsageMetrics collects the CPU usage of each thread.
+func updateThreadCPUUsageMetrics() []prometheus.Metric {
+	cpuTimeStat.RLock()
+	defer cpuTimeStat.RUnlock()
+	threadCPUUsage := make([]prometheus.Metric, cpuNumber)
+	for i := 0; i < cpuNumber; i++ {
+		threadCPUUsage[i] = prometheus.MustNewConstMetric(
+			descriptors[threadCPUUsageDesc],
+			prometheus.GaugeValue,
+			cpuTimeStat.CPUUsage[i],
+			strconv.Itoa(i),
+		)
+	}
+	return threadCPUUsage
+}
+
+// updatePoolCPUUsageMetrics collects the CPU usage of pools defined by podpools-policy.
+func updatePoolCPUUsageMetrics(ppm *Metrics) ([]prometheus.Metric, error) {
+	if ppm == nil {
+		return nil, fmt.Errorf("Podpools metrics used to count pool CPU usage is missing.")
+	}
+	// Sort the pool metrics.
+	poolNames := make([]string, 0, len(ppm.PoolMetrics))
+	for poolName := range ppm.PoolMetrics {
+		poolNames = append(poolNames, poolName)
+	}
+	sort.Sort(sort.StringSlice(poolNames))
+
+	// Calculate the CPU usage of a pool and send to prometheus.
+	poolCPUUsageMetrics := make([]prometheus.Metric, len(poolNames))
+	poolCPUUsageList := make(map[string]float64, len(poolNames))
+	cpuTimeStat.RLock()
+	defer cpuTimeStat.RUnlock()
+	for index, poolName := range poolNames {
+		cpus := resolvePools(ppm.PoolMetrics[poolName].CPUs)
```

Comment:
Currently what happens on lines 195..200 are the last steps of conversion chain:
1. value in `pool.CPUs` (cpuset) converted by `cpuset.String()` (on line 107) to...
2. value in `PoolMetrics[poolName].CPUs` (string) converted by `resolvePools` to...
3. value in `cpus` (string) converted by `Split` to...
4. value in `cpusPerPoolList` ([]string), elements converted by `Atoi` to...
5. cpuIdInt (int)

I would suggest simplifying this structure by adding line 107 and storing `[]int` to a new attribute:
```
...
policyMetrics.PoolMetrics[pool.PrettyName()].CPUs = pool.CPUs.String()
policyMetrics.PoolMetrics[pool.PrettyName()].CPUIds= pool.CPUs.ToSlice()
...
```

Then, you can change the for loop on line 199 to the form
```
for cpuId := range ppm.PoolMetrics[poolName].CPUIds {
```
...and get rid of the `resolvePools` function and this chain of conversions.